### PR TITLE
Training scripts, VAE, and MINE convenience code

### PR DESCRIPTION
**Pull request by AnejSvete**
This PR implements 2 scripts for end-to-end training of `MINE` on synthetic data and on VAE representations of `dSprites`.

It also optimises some code structure for generating the data and for manipulating datasets.

Yeah, there is quite a bit of new code and some refactoring of the existing one...

There are two open questions from my side:
1. I could not come up with a good way to structure the various util methods and files, so I ended up with many `utils` files, which I think could be structured better.
2. More importantly, despite setting the seed (globally?) with `pl.seed_everything` in the training script (`dsprites_vae.py`), the training does not seem to be deterministic. I am not sure how to figure out the cause and how to fix it.

### CODE REVIEW

File: latte/dataset/dsprites.py [hunk 1]
```diff
@@ -199,30 +209,35 @@ def decompose(X: np.ndarray, method: str, n_components: int) -> DecompositionRes
 
     X = StandardScaler().fit_transform(X)
     X = trans.fit_transform(X)
+    # X = StandardScaler().fit_transform(X)
```
Comment (pawel-czyz) [inline latte/dataset/dsprites.py:212]: I'd just remove that :slightly_smiling_face:

File: latte/dataset/utils.py [hunk 2]
```diff
@@ -84,3 +86,32 @@ def download(target: Union[str, pathlib.Path], _source: str, _checksum: str) ->
             "If it doesn't help, check whether the MD5 checksum are right. "
             "Note that opening a corrupted file may be a threat to your computer, due to pickle."
         )
+
+
+def split(
```
Comment (pawel-czyz) [inline latte/dataset/utils.py:91]: Yes, this works :slightly_smiling_face: 

In the future, perhaps [this function](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) will be useful.

File: latte/mine/estimation.py [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+        learning_rate: The learning_rate of the MINE model
+        manifold_learning_rate: The learning rate for the k-frame manifold optimiser
+        alpha: The bias-compensating alpha value for the MINE estimator
+
+    Returns:
+        The constructed model
+    """
+    if manifold:
+        assert subspace_size != 0, "No target subspace size provided"
+        S = mine.ManifoldStatisticsNetwork(
```
Comment (pawel-czyz) [inline latte/mine/estimation.py:64]: Possible todo item: generalizing this so that we can provide a non-hardcoded network for MINE.

File: latte/models/datamodules.py [hunk 4]
```diff
@@ -100,55 +150,50 @@ def __init__(
...
-        self.mine_test = ds.MINEDataset(X=torch.from_numpy(self.X_test), Z=torch.from_numpy(self.y_test))
 
-    def train_dataloader(self):
-        return DataLoader(
-            self.mine_train, batch_size=self.batch_size, num_workers=self.num_workers, persistent_workers=True
-        )
+        def _transform(img):
+            pepper = self.rng.binomial(1, p=self.noise_amount, size=img.shape)
+            salt = self.rng.binomial(1, p=self.noise_amount, size=img.shape)
+            return torch.from_numpy(np.clip(img + salt - pepper, 0, 1)).type(torch.FloatTensor)
```
Comment (pawel-czyz) [inline latte/models/datamodules.py:189]: Nice!

File: latte/utils/utils.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+from typing import List
+from collections import defaultdict as dd
```
Comment (pawel-czyz) [inline latte/utils/utils.py:2]: I'm a big fan of using explicit names, i.e., `defaultdict` is easier to understand than `dd` (I needed to go back to the header to check it) :slightly_smiling_face:

File: setup.cfg [hunk 6]
```diff
@@ -11,6 +11,7 @@ long_description_content_type = text/markdown
 packages = latte
 python requires = >= 3.8
 install_requires =
+    coloredlogs
```
Comment (pawel-czyz) [inline setup.cfg:14]: Cool, I learned something new! :slightly_smiling_face:

**Review by pawel-czyz — APPROVED** [review]
Nice! :slightly_smiling_face:

**Comment by AnejSvete** [comment]
Pinning `attrs` to the version I have locally seems to have solved it.
